File: libs/moqtail-ts/src/model/control/publish_namespace_error.ts
import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { ReasonPhrase } from '../common/reason_phrase'
import { LengthExceedsMaxError } from '../error/error'
import { PublishNamespaceErrorCode, ControlMessageType, publishNamespaceErrorCodeFromBigInt } from './constant'

export class PublishNamespaceError {
  constructor(
    public readonly requestId: bigint,
    public readonly errorCode: PublishNamespaceErrorCode,
    public readonly reasonPhrase: ReasonPhrase,
  ) {}

  getType(): ControlMessageType {
    return ControlMessageType.PublishNamespaceError
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.PublishNamespaceError)
    const payload = new ByteBuffer()
    payload.putVI(this.requestId)
    payload.putVI(this.errorCode)
    payload.putReasonPhrase(this.reasonPhrase)
    const payloadBytes = payload.toUint8Array()
    if (payloadBytes.length > 0xffff) {
      throw new LengthExceedsMaxError(
        'PublishNamespaceError::serialize(payloadBytes.length)',
        0xffff,
        payloadBytes.length,
      )
    }
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)
    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): PublishNamespaceError {
    const requestId = buf.getVI()
    const errorCodeRaw = buf.getVI()
    const errorCode = publishNamespaceErrorCodeFromBigInt(errorCodeRaw)
    const reasonPhrase = buf.getReasonPhrase()
    return new PublishNamespaceError(requestId, errorCode, reasonPhrase)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest

  describe('PublishNamespaceError', () => {
    test('roundtrip', () => {
      const requestId = 12345n
      const errorCode = PublishNamespaceErrorCode.ExpiredAuthToken
      const reasonPhrase = new ReasonPhrase('tis I sir lancelot of camelot')
      const announceError = new PublishNamespaceError(requestId, errorCode, reasonPhrase)
      const serialized = announceError.serialize().toUint8Array()
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.PublishNamespaceError))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = PublishNamespaceError.parsePayload(frozen)
      expect(deserialized.requestId).toBe(announceError.requestId)
      expect(deserialized.errorCode).toBe(announceError.errorCode)
      expect(deserialized.reasonPhrase.phrase).toBe(announceError.reasonPhrase.phrase)
      expect(frozen.remaining).toBe(0)
    })

    test('excess roundtrip', () => {
      const requestId = 67890n
      const errorCode = PublishNamespaceErrorCode.InternalError
      const reasonPhrase = new ReasonPhrase('wake me up')
      const announceError = new PublishNamespaceError(requestId, errorCode, reasonPhrase)
      const serialized = announceError.serialize().toUint8Array()
      const excess = new Uint8Array([9, 1, 1])
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.PublishNamespaceError))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining - 3)
      const deserialized = PublishNamespaceError.parsePayload(frozen)
      expect(deserialized.requestId).toBe(announceError.requestId)
      expect(deserialized.errorCode).toBe(announceError.errorCode)
      expect(deserialized.reasonPhrase.phrase).toBe(announceError.reasonPhrase.phrase)
      expect(frozen.remaining).toBe(3)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })

    test('partial message', () => {
      const requestId = 112233n
      const errorCode = PublishNamespaceErrorCode.MalformedAuthToken
      const reasonPhrase = new ReasonPhrase('Uvuvwevwevwe')
      const announceError = new PublishNamespaceError(requestId, errorCode, reasonPhrase)
      const serialized = announceError.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const frozen = new FrozenByteBuffer(partial)
      expect(() => {
        frozen.getVI()
        frozen.getU16()
        PublishNamespaceError.parsePayload(frozen)
      }).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/control/publish_namespace_error.rs
use bytes::{BufMut, Bytes, BytesMut};

use super::constant::PublishNamespaceErrorCode;

use crate::model::common::reason_phrase::ReasonPhrase;

use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::control::constant::ControlMessageType;
use crate::model::control::control_message::ControlMessageTrait;
use crate::model::error::ParseError;

#[derive(Debug, Clone, PartialEq)]

pub struct PublishNamespaceError {
  pub request_id: u64,
  pub error_code: PublishNamespaceErrorCode,
  pub reason_phrase: ReasonPhrase,
}
impl PublishNamespaceError {
  pub fn new(
    request_id: u64,
    error_code: PublishNamespaceErrorCode,
    reason_phrase: ReasonPhrase,
  ) -> Self {
    PublishNamespaceError {
      request_id,
      error_code,
      reason_phrase,
    }
  }
}

impl ControlMessageTrait for PublishNamespaceError {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::PublishNamespaceError)?;

    let mut payload = BytesMut::new();
    payload.put_vi(self.request_id)?;
    payload.put_vi(self.error_code)?;
    payload.extend_from_slice(&self.reason_phrase.serialize()?);

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "PublishNamespaceError::serialize(payload_length)",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;
    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);

    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let request_id = payload.get_vi()?;

    let error_code_raw = payload.get_vi()?;
    let error_code = PublishNamespaceErrorCode::try_from(error_code_raw)?;

    let reason_phrase = ReasonPhrase::deserialize(payload)?;

    Ok(Box::new(PublishNamespaceError {
      request_id,
      error_code,
      reason_phrase,
    }))
  }

  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::PublishNamespaceError
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let request_id = 12345;
    let error_code = PublishNamespaceErrorCode::ExpiredAuthToken;
    let reason_phrase = ReasonPhrase::try_new("tis I sir lancelot of camelot".to_string()).unwrap();
    let announce_error = PublishNamespaceError::new(request_id, error_code, reason_phrase.clone());

    let mut buf = announce_error.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::PublishNamespaceError as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = PublishNamespaceError::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, announce_error);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let request_id = 67890;
    let error_code = PublishNamespaceErrorCode::InternalError;
    let reason_phrase = ReasonPhrase::try_new("wake me up".to_string()).unwrap();
    let announce_error = PublishNamespaceError::new(request_id, error_code, reason_phrase.clone());

    let serialized = announce_error.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();

    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::PublishNamespaceError as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining() - 3);
    let deserialized = PublishNamespaceError::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, announce_error);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let request_id = 112233;
    let error_code = PublishNamespaceErrorCode::MalformedAuthToken;
    let reason_phrase = ReasonPhrase::try_new("Uvuvwevwevwe".to_string()).unwrap();
    let announce_error = PublishNamespaceError::new(request_id, error_code, reason_phrase.clone());

    let mut buf = announce_error.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::PublishNamespaceError as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());

    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = PublishNamespaceError::parse_payload(&mut partial);
    assert!(deserialized.is_err());
  }
}
